server/src/main/java/io/crate/execution/dml/ArrayIndexer.java

```diff
@@ -201,7 +201,7 @@ private void handleNullArrayUpcast(List<?> values,
         StorageSupport<?> storageSupport = type.storageSupport();
         assert storageSupport != null;  // null storage will have thrown an exception in buildReference
         ValueIndexer<Object> indexer
-            = (ValueIndexer<Object>) storageSupport.valueIndexer(ref.ident().tableIdent(), ref, _ -> ref);
+            = (ValueIndexer<Object>) storageSupport.valueIndexer(ref.ident().tableIdent(), ref, _ -> null);
```

Comment:
Why is the change here?

Reply:
The original code was wrong as it would always return the top-level `array_of_object` reference on every `getRef` call ignoring the given `ColumnIdent`. Such, processing e.g. resolving the reference for the child ident `arr['x']` would return the `arr` reference. Passing `NULL` here, will bypass it due to https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/ObjectIndexer.java#L86. 

Creating a reference resolver on the fly here would be maybe the best solution, but I'm not really sure if this is worth the effort.

Reply:
Ah no, the `null` handling at the `ObjectIndexer` is exactly for this case were inner types aren't known and will be added later on to the schema.
So I think, passing in `null` here is correct.

Reply:
I've checked how this behaves at `master`, and there the `ObjectType` passed in to the `ObjectIndexer` does not contain any inner types (it's `ObjectType.UNDEFINED`) and such the passed in (wrong) `getRef` is never called.